### DISCUSSION TITLE
How to help get non_blocking across the finish line?

### DISCUSSION THREAD

**Discussion by hugetim** [anvilistas/anvil-labs#70] [General]
With reference to #52, I'd also love to use async server calls in a production-ish app. I could just copy the code, but if I could spend a similar amount of time helping it along to become part of anvil-extras, I'd rather do that.

The thing that strikes me as the most likely obstacle is that the API seems tentative in places. (For example, the docs note how `Interval` works different from an Anvil Timer component in a way that could trip people up.)

Here are some suggested changes for discussion--but no need to respond to any of these. Rather I presume each of them has already been rejected for good reason. But feel free to adopt or discuss further if any of them do seem helpful:

1. Rename `call_server_async` to `server_call_async` to better mirror Anvil's (and kompot's) API.
2. Change `Interval` so that setting `delay=0` stops it (the same as `delay=None`).
3. Rename `Interval`'s `delay` attribute to `period'.
4. Rename `Timeout` to `Delay` (and rename its attribute `delay` to `period`). (With `period=0`, it should execute without delay.)
5. Rename the `clear` methods to `stop`.

Also, more generally, I may not know enough to ask this question intelligibly, but would it be better/possible to implement async stuff in Skulpt directly, like using the async/await keywords? (Really, my level of understanding on this is just, "I hear people in Python podcasts talking about async/await, so maybe that's relevant.") In other words, to implement the relevant Python syntax/standard library rather than adding a new package? But I guess that may not be ideal because of some differences between JS and Python that I don't fully understand?

Another way I might be able to contribute is trying to add cases to the [existing tests](https://github.com/anvilistas/anvil-labs/blob/7ff3eb8b1a067e362d05f7a50acb214d95520737/client_code/non_blocking.py#L217)? Would that be helpful?

**Comment by s-cork** [comment]
Re skulpt. Unfortunately I think that'll take a long time.

My main problem with async await in Python for anvil is that you always have to block somewhere to run the event loop. Async calls then do their thing. But I've never seen async code that gives a callback and says 'give me the result when you're done, and I'll continue doing stuff until then'. 

It's more 'give me the result when you're done and I guess I'll just wait patiently until then'. 

I'd love to see an example of the former. But that seems to be more of a threading pattern. 

The tasks in the event loop are free to hand over to each other while they await. But you always seem to have to wait for all tasks to complete before moving on in your code.

So I've never quite gotten my head around how it would look in Anvil. I haven't done that much serious python async await stuff though. 

___

I was never fully happy with the API so I'll have a think about your suggestions. I was quite influenced by the JavaScript api. I did think about delay=0. 

@meatballs did you use either of these in your code yet?

How do we feel about the rest of the API?

**Reply by meatballs** [reply]
Yes, I've used them but, at the moment, I maintain my own copy of the app and depend on that, so no problem changing things.

**Reply by hugetim** [reply]
(Feel free to disregard--this is the blue sky thinking of a complete noob on all things async.)

On Python async, I found these helpful just now:
- https://stackoverflow.com/questions/68139555/difference-between-async-await-in-python-vs-javascript
- https://stackoverflow.com/questions/58261157/how-to-have-something-similar-to-javascript-callbacks-in-python

It sounds to me like it would be possible to reflect in Python the async nature of how certain bits of the Anvil client-side already work, but that would require spelling out explicitly in Python the sort of "meta" Anvil client side. For instance, I have some intuition from experience of how click events and Timer tick events get handled in Anvil, and I can now make sense of that in terms of an event loop. And I can also now better comprehend what might be going on in JS under the hood when I attempt to raise an Anvil `alert` from inside another alert (that is, it queues an event--showing the new alert--that will run once it's done waiting for the current alert to be closed). It seems like it should be possible to spell out in Python terms how those things are working, that there's essentially an event loop begun on app startup, with everything else running in that context.

If I'm understanding correctly, once the entire Anvil client side is understood as running in a Python event loop, you should be able to do async/await stuff akin to how those work in JavaScript.

**Reply by s-cork** [reply]
Yes - Skulpt could implement a version of async await, which assumes the event loop is already happening.
But it couldn't really match Python's implementation exactly since the event loops is already going on and can't be stopped.

Some other notes:
When we do sleep in skulpt - it's more akin to asyncio.sleep
A sleep in skulpt doesn't block, it gives the Javascript event loop the opportunity to execute some other code.
(you'll see I've suggested `sleep(0)` a number of times in recent posts)

When we do `anvil.server.call()` it's more like `await anvil.server.call()`
Other code can execute while the call is happening since control of the event loop can be given to other tasks.

You might find Brython's async implementation docs worth reading:
https://brython.info/static_doc/en/aio.html

**Reply by s-cork** [reply]
> Yes, I've used them

@meatballs have you used them both?
How did you find the API?

---

Just throwing an idea out there - we could literally wrap the javascript api
```python
self.handle_id = non_blocking.set_timeout(self.my_func, 1)
...
non_blocking.clear_timeout(self.handle_id)

```

The `set_timeout` and `set_interval` return call ids that you can clear before (or after) the function is executed.
There's something nice about this API.
I was originally trying to make the API feel more like Anvil's Timer API.

Of course if those words don't make sense we could adjust:

```python
handle_id = non_blocking.delay(some_func, 1)
handle_id = non_blocking.repeat(some_func, 1)
...
non_blocking.clear(handle_id)
# or
non_blocking.end(handle_id)

```


The heartbeat example would something like

```python

from anvil_labs import non_blocking

i = 0
def do_heartbeat():
    global heartbeat_id, i
    if i >= 42:
        non_blocking.clear(heartbeat_id)
    print("da dum")
    i += 1

heartbeat_id = non_blocking.repeat(do_heartbeat, 1)

```


```python
    from anvil_labs import non_blocking

    pending = []

    def do_save():
        global pending
        pending, saves = [], pending
        if not saves:
            return
        anvil.server.call_s("save", saves)

    save_timeout_id = None

    def on_save(saves):
        global pending, save_timeout_id
        non_blocking.clear(save_timeout_id)
        pending.extend(saves)
        save_timeout_id = non_blocking.delay(do_save, 1)
```

We could also make the handle_id a more opaque object.
Something that has a single method, `clear` or `end`.

like

```python

from anvil_labs import non_blocking

i = 0
def do_heartbeat():
    global heartbeat_ref, i
    if i >= 42:
        heartbeat_repeat_ref.clear() # or .end() or .cancel() or .stop()
    print("da dum")
    i += 1

heartbeat_ref = non_blocking.repeat(do_heartbeat, 1)

```

**Reply by meatballs** [reply]
I do find the current api tricky - it's fine to use but when I come back to the code later, it's not immediately obvious how it works.

I like the idea of wrapping the js api but with changed terms. `repeat` sounds great, but I'm not sold on `delay`. How about `defer` instead?

**Reply by s-cork** [reply]
yeah I could get behind `defer`

**Reply by hugetim** [reply]
Actually, I kind of like just keeping the JavaScript terms now that I understand where they come from. If the API is going to closely parallel the JavaScript, it might be helpful to keep the names the same as well. But I'm fine either way. (Regardless, no need to worry about `set_interval(some_func, 0)` behaving differently from an Anvil Timer with interval 0 anymore.)

**Reply by s-cork** [reply]
that's a 🔧

We currently have
```python
from anvil_labs import non_blocking

i = 0
def do_heartbeat():
    global heartbeat_ref, i
    if i >= 42:
        heartbeat_ref.cancel()
    print("da dum")
    i += 1

heartbeat_ref = non_blocking.repeat(do_heartbeat, 1)
```

We could also have the same code but with the last line as:

```python
heartbeat_ref = non_blocking.set_interval(do_heartbeat, 1)
```

I worry that the javascript API is fine if you know its origins.
But if you don't it's not that obvious.
Hard for me to see it with fresh eyes though 👀


Worth noting that anyone can always just roll your own version

```python
from anvil.js.window import setInterval
ref = setInterval(fn, 1000)
```

**Reply by hugetim** [reply]
Yeah, if the JavaScript `setInterval` is that easily accessible, maybe it's better to just direct people to that for these functions rather than create thin Python wrappers?

On balance, I do like the `repeat` and `defer` non_blocking functions, though. They will be appreciated by people like me coming without that JavaScript background. (Apologies for the indecision.)

**Comment by hugetim** [comment]
@s-cork from the other comment thread:
> When we do `anvil.server.call()` it's more like `await anvil.server.call()`
> Other code can execute while the call is happening since control of the event loop can be given to other tasks.

Whoa, this is not what I expected. So that means a Timer tick event could fire (and execute client code) while a server function is running?

But user input via the UI is blocked during a server call (even `call_s`), right?

**Reply by s-cork** [reply]
Try it - run a server call and start clicking buttons
The button should have a print statement.
Add a sleep call in the server function so that it lasts a while...

**Reply by s-cork** [reply]
And Timer - yeah they're are just glorified `setTimeout` calls that reset the timeout to the current `.interval` value after each call
So a timer tick event can also execute during a server call

You can see in the source code for the Timer
https://github.com/anvil-works/anvil-runtime/blob/89b6d15bc5a1bd2bc9aee4938abd3bbf26b6e0ec/client/js/components/Timer.js#L74-L85

**Reply by hugetim** [reply]
Ok, I see. Thanks!
But the server is strictly serial, right? Even when used with call_async, server calls are going to be queued on the server in the order they were called and run one-after-the-other.

**Reply by s-cork** [reply]
yes that's right - they will execute in the order they are called.
And one server call won't run until the previous one has finished.

**Reply by hugetim** [reply]
Also, is there some other reason why a button would sometimes not register a click? I thought that had happened to me sometimes, and I attributed it to a Timer running a server.call_s, but maybe I was just imagining things (and just hadn't fully pressed down the mouse button or something).

**Reply by s-cork** [reply]
not that I know of

**Comment by hugetim** [comment]
(In reference to: https://github.com/anvilistas/anvil-labs/discussions/70#discussioncomment-2646755)
Ok, I think what must have happened is that I was expecting the button to
trigger a server call with the spinner. But the silent server call was
still running. So the button click function ran, but it stopped at the
"await server.call" line until the silent server call was finished. If
that's right (I won't be able to test it for a bit), I may make an Anvil
feature request to start the spinner right away, even if the server call
hasn't started (been sent?) yet due to a backup in the server queue.

**Reply by s-cork** [reply]
Ah yes I could how that might happen. 

A repo app and a forum post will be the way to go 👍

**Reply by hugetim** [reply]
Nope, I was wrong. The spinner starts right away, even if there's already a silent server call running.

**Comment by hugetim** [comment]
For what it's worth, now that I understand that ordinary `call_s` server calls do not stop other client code from functioning (with the only exception being code directly following the `call_s` in the specific function it is a part of), I don't currently have a use for the functionality provided by the NonBlocking module.
